### ISSUE TITLE
[Question] How does the producer know if it loses connection from the server

### ISSUE THREAD

**Issue by mobilotronic** [tulios/kafkajs#1075]
Hi, first thanks for making this library available :)

My test environment looks like this:
- KafkaJS version is 1.15.0
- Kafka server is running locally in docker
- A very simple nodeJS app creates a Producer and connects to the server successfully.
- The kafka/zookeeper pods are stopped.

The disconnect event of the producer never fires.
If logging is set to DEBUG, you can see that the socket's "end" event fires, but that event is not "bubbled" up all the way to the producer.

The question is this: How does the producer know that the underlying connection to Kafka is lost?

Is there another way to monitor the status of the underlying connection?

Thanks.

**Comment by Nevon** [comment]
This comes up quite often, and my opinion is that users shouldn't worry about internal things like what brokers a producer is connected to, but rather look at the high level behavior of the client.

It is perfectly normal and acceptable that a producer doesn't have a constant connection to any broker. When the producer is not actively sending anything, there's no reason it should keep a connection alive. If the connection is idle, it will eventually be closed. When the producer is then used in order to send a message, it will reconnect. It's really no different from if the producer is connected to broker A, but needs to produce a message to a partition for which broker B is the leader - it just connects to broker B in order to send the message.

We do emit events for connect and disconnect, but they are only half-true, since a connection may be closed from the broker side without us knowing it if it was closed without sending a FIN packet. https://kafka.js.org/docs/instrumentation-events#a-name-producer-a-producer  In fact, looking at the implementation, the only time we emit the DISCONNECT event is when the producer is explicitly disconnected (i.e. you called `producer.disconnect`). 

In summary, I would question the need to monitor the status of the underlying _connections_. Monitor the outcomes instead. Is `producer.send` rejecting? Then you have a problem. If it's not rejecting, you don't have a problem.

**Comment by mobilotronic** [comment]
Nevon, appreciate you taking the time to reply.

This is what I am currently doing, monitoring the outcomes.

I guess I was sidetrack/confused a bit on why those events were not firing.

I suppose I was also looking for a more "straight forward" (which is a matter of perspective), to get connection state of the producer to the server.
Say for example the connection state (connected/disconnected) was part of a `readyness` probe in a container orchestration tool such as k8s.

What I am doing currently seems to be the most accurate way to get/infer the connection status.

Thanks.
